### DISCUSSION TITLE
🗺️ Roadmap 3.0

### DISCUSSION THREAD

**Discussion by kamilkisiela** [kamilkisiela/graphql-inspector#1371] [General] [answered]
Track current progress in ["Full Product" project](https://github.com/kamilkisiela/graphql-inspector/projects/4).

# Free and OpenSource

As always, all the things done in GraphQL Inspector are free to use and open-sourced. Everything can be self-hosted. That's the core value of Inspector.

- [Granular CLI](#granular-cli)
- [Change Log, Versioning and History of your GraphQL API](#changelog)
- [Stable GitHub Action and App with few more commands](#stable-github)
- [Track and analyze live usage](#live)
- [Notify about changes in Schema](#notifications)
- [Scheduled removal of deprecated fields](#scheduled-removal)
- [More advanced validation of operations and fragments](#validation)
- [User Interface and GraphQL API](#ui-api)
- [Closer to Teams and your Company structure](#teams)

<a name="granular-cli"></a>
## Granular CLI

**Related issue** #1201

- Each command as a separate package
- Each loader as a separate package
- Config file to define used commands and loaders
- Config file to define actions like diff, validate etc, with an ID so in CI you do: `graphql-inspector run <id>` and it does set of commands

**Loader** is a package that is responsible for loading Schemas and Documents from different kind of sources. Right now `@graphql-inspector/load` has all the loaders, it can load pretty much anything. I want to split it into packages with single responsibility. 
For example, if you want to "teach" GraphQL Inspector how to load GraphQL Schema from multiple `.graphql` files, you use `@graphql-inspector/graphql-loader` but if you want to use a running endpoint you choose `@graphql-inspector/url-loader`. You define those loaders in a config file.

**Command** is a package with a single responsibility. Current CLI has `diff`, `validate` and other commands but maybe you don't use them all. That's why I want to put each command in a separate package. Just like loaders.

Config file would look like this:

```js
module.exports = {
  use: {
    loaders: ['url', 'graphql'],
    commands: ['diff', 'introspect'],
  },
};
```

<a name="changelog"></a>
## Change Log, Versioning and History of your GraphQL API

**Related issue** #213

We want to let people use versioning and create a nice looking Change Log.
This is going to be something that everyone can deploy in their infrastructure or use our free service.

<a name="stable-github"></a>
## Stable GitHub Action and App with few more commands

**Related issues** #106 #107 #267

Both things were quite experimental but I want to focus on them more and stabilize the API. This will include a better setup in the monorepo + e2e tests (example repositories).

<a name="live"></a>
## Track and analyze live usage

**Related issues** #212 #444

This is something I've been working on for a long time. I want to enable you to track and analyze how your GraphQL API is consumed. Which part of schema is used, how often, what's the performance and so on.
It will be integrated with Github Action, Github App, CLI and a service. How? An example, you deprecate a piece of schema and after few weeks you want to remove it. With collected data you can be sure it won't break your existing API consumers. There are more use cases, it has an endless potential.

<a name="notifications"></a>
## Notify about changes in Schema

**Related issue** #214

Whenever someone makes a change in a schema, you or your team will get a notification. It can happen on Slack, Discord or even through a Webhook. More on teams later...
You will be able to track both Pull Requests and your master branch.

<a name="scheduled-removal"></a>
## Scheduled removal of deprecated fields

**Related issue** #215

When a field is marked as deprecated, you can set a date when it will be removed. Since GraphQL Inspector understand which part of schema belongs to which team, all the affected developers will get a notification (once the field is deprecated and few more times until it is officially removed).

<a name="validation"></a>
## More advanced validation of operations and fragments

**Related issue** #1

I want to let people set a threshold when an operation becomes too complex or too deeply nested. Let people write their own validation rules. Integrate a Linter or create a new one.

<a name="ui-api"></a>
## User Interface and GraphQL API

**Related issue** #12

This one is quite interesting and it involves:
 - desktop app (electron or served webapp via CLI)
 - web app (with mobile version)
 - GraphQL API

This is something I always wanted to do... An application that can be deployed, used as a service or ran straight from the CLI.
This application can explore the GraphQL Schema and all Operations and Fragments in your project (with links to files in your codebase).

You will be able to validate documents, compare schemas, see the live usage of GraphQL API and pretty much whatever GraphQL Inspector lets you do in CLI or Actions.

Because we create UI and API in separation, it lets us share the UI on web app and desktop app and consume the same API. It's also possible to expose that API and use just that, directly or for intergration with other tools.

<a name="teams"></a>
## Closer to Teams and your Company structure

I want to develop an easy to use tool that matches how structure of your organization. You probably work in teams and those teams own parts of schema. GraphQL Inspector should understand it and easier your day to day workflow of working with GraphQL.

**Comment by mesca216** [comment]
Thanks for all guys!!! Really

**Comment by yenda** [comment]
Really great tool, thanks for all the work!

We have different graphql schemas in our monorepo, it would be great to be able to specify the schema path in the env as well so that we can have the Github App check and notify changes in all schemas.

**Comment by kamilkisiela** [comment, marked as answer]
Some of those features are now implemented in [GraphQL Hive](https://graphql-hive.com) :)
